tsl/src/remote/connection.c

```diff
@@ -890,30 +919,197 @@ remote_connection_get_result_error(const PGresult *res, TSConnectionError *err)
 	fill_result_error(err, ERRCODE_CONNECTION_EXCEPTION, "", res);
 }
 
+static bool
+wait_and_consume_input(const TSConnection *conn, long timeout, uint32 events)
+{
+	WaitEvent event;
+
+	ModifyWaitEvent(conn->wes, conn->sockeventpos, events, NULL);
+	WaitEventSetWait(conn->wes, timeout, &event, 1, PG_WAIT_EXTENSION);
+
+	if (event.events & WL_LATCH_SET)
+	{
+		ResetLatch(MyLatch);
+		CHECK_FOR_INTERRUPTS();
+	}
+	if (event.events & WL_SOCKET_READABLE)
+	{
+		if (PQconsumeInput(conn->pg_conn) == 0)
+			return false;
+	}
+
+	/* Other events, just return */
+	return true;
+}
+
+PGresult *
+remote_connection_get_result(const TSConnection *conn)
+{
+	PGresult *pgres = NULL;
+	int ret = 1;
+
+	do
+	{
+		ret = PQisBusy(conn->pg_conn);
+
+		switch (ret)
+		{
+			case 1: /* PQgetResult would block */
+				if (!wait_and_consume_input(conn, -1, WL_SOCKET_READABLE))
+					return NULL;
+				break;
+			case 0: /* PQgetResult would not block */
+				pgres = PQgetResult(conn->pg_conn);
+				break;
+			default:
+				pg_unreachable();
+				break;
+		}
+	} while (ret != 0);
+
+	return pgres;
+}
+
+/*
+ * Flush a connection after writing data on a non-blocking socket.
+ *
+ * Implements non-blocking flush according to the documentation for PQflush(),
+ * but also observes PostgreSQL interrupts.
+ */
+bool
+remote_connection_flush(const TSConnection *conn, TSConnectionError *err)
+{
+	int ret = 1;
+
+	do
+	{
+		ret = PQflush(conn->pg_conn);
+
+		if (ret == 1)
+		{
+			if (!wait_and_consume_input(conn, -1, WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE))
```

Comment:
Yes, this is the documentation for `PQflush`: 


After sending any command or data on a nonblocking connection, call [PQflush](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQFLUSH). If it returns 1, wait for the socket to become read- or write-ready. If it becomes write-ready, call [PQflush](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQFLUSH) again. If it becomes read-ready, call [PQconsumeInput](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQCONSUMEINPUT) , then call [PQflush](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQFLUSH) again. Repeat until [PQflush](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQFLUSH) returns 0. (It is necessary to check for read-ready and drain the input with [PQconsumeInput](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQCONSUMEINPUT) , because the server can block trying to send us data, e.g., NOTICE messages, and won't read our data until we read its.) Once [PQflush](https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQFLUSH) returns 0, wait for the socket to be read-ready and then read the response as described above.